generator/internal/parser/httprule/http_rule_parser_test.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httprule
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"testing"
+)
+
+func TestProtobuf_Parse(t *testing.T) {
+	tests := []struct {
+		path        string
+		want        *PathTemplate
+		explanation string
+	}{
+		{"/foo", template(literal("foo")), ""},
+		{"/foo/bar", template(literal("foo"), literal("bar")), ""},
+		{"/v1/*/foo", template(literal("v1"), match(), literal("foo")), ""},
+		{"/v1/**/foo", template(literal("v1"), matchR(), literal("foo")), ""},
+		{"/foo:bar", templateV(segments(literal("foo")), verb("bar")), ""},
+		{"/foo/{bar}", template(literal("foo"), varr("bar")), ""},
+		{"/foo/{bar.baz}", template(literal("foo"), varr("bar", "baz")), ""},
+		{"/foo/{bar=baz}", template(literal("foo"), varrs(ids("bar"), segments(literal("baz")))), ""},
+		{"/foo/{bar=*}", template(literal("foo"), varrs(ids("bar"), segments(match()))), ""},
+		{"/foo/{bar=*}/baz", template(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), ""},
+		{"/foo/{bar=*}/baz:qux", templateV(segments(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), verb("qux")), ""},
+		{"foo", nil, "path must start with slash"},
+		{"/", nil, "path cannot end with slash"},
+		{"/foo/", nil, "path cannot end with slash"},
+
+		// the following test is failing because * is a sub-delimiter, which is allowed in the LITERAL segment
+		//{"/foo/***/bar", nil, "wildcard literal cannot exceed two *"},
+
+		{"/%0f", template(literal("%0f")), ""},
+		{"/%0z", nil, "bad percent encoding"},
+		{"/foo//bar", nil, "segment is too short"},
+		{"/foo/:", nil, "verb is too short"},
+		{"/foo/{}/bar", nil, "var too short"},
+		{"/foo/{a.}/bar", nil, "var identifier too short"},
+		{"/foo/{.a}/bar", nil, "var identifier too short"},
+		{"/foo/{a=}/bar", nil, "var value too short"},
+		{"/foo/{9bar}", nil, "var identifier has bad first character"},
+		{"/foo/{bar9}", template(literal("foo"), varr("bar9")), ""},
+		{"/foo/{b&r}", nil, "var identifier has bad character"},
+		{"/foo/:bar", nil, "verb cannot come after slash"},
+		{"/foo:bar/baz", nil, "verb must be the last segment"},
```

Comment:
This may actually be a bug. According to the spec, `:` is a pchar, and valid as a path segment. I am not sure if it is realistic to expect it to be used that way in this use case.

Reply:
If you meant "a bug in the definition of http rules" then I agree.  It is possible that the `literals` should be more restrictive, maybe "only things that appear in resource names" though I am not sure if that is defined anywhere?

Reply:
I think it's a bug in my implementation of the RFC 3986. But I also think that our interpretation of the http rules to include all of `pchar` as a valid `LITERAL` may be too broad. 
I am inclined, for now, to remove these from the parser `sub-delims | ":" | "@"` and leave them commented out with a pointer to this discussion, in case we need to revert this in the future. 
WDYT?

Reply:
SGTM.

Reply:
Done! Updated the documentation in the parser to include the more restrictive parsing logic.